kolibri/auth/test/test_permissions.py

```diff
@@ -1,454 +1,568 @@
-#         self.assertFalse(self.coach1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[1]))
-
-#     def test_remove_learner_specific_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[0]))
-
-#     def test_remove_learner_rejects_non_classroom_objects(self):
-#         with self.assertRaises(InvalidPermission):
-#             self.admin.has_perm(self.AUTH_REMOVE_LEARNER, obj={})
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.add_facility_admin     ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_ADD_FACILITY_ADMIN = 'auth.add_facility_admin'
-
-#     def test_add_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.remove_facility_admin  ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_REMOVE_FACILITY_ADMIN = 'auth.remove_facility_admin'
-
-#     def test_remove_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
+"""
+Tests of the permissions on specific models in the auth app. For tests of the permissions system itself, see test_permission_classes.py
+"""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from django.test import TestCase
+
+from .helpers import create_dummy_facility_data
+
+from ..constants import role_kinds
+from ..models import DeviceOwner, Facility, Classroom, LearnerGroup, Role, FacilityUser
+
+
+class FacilityPermissionsTestCase(TestCase):
+    """
+    Tests of permissions for reading/modifying Facility instances
+    """
+
+    def setUp(self):
+        self.data1 = create_dummy_facility_data()
+        self.data2 = create_dummy_facility_data()
+        self.device_owner = DeviceOwner.objects.create(username="boss")
+
+    def test_facility_users_cannot_create_facility(self):
+        """ FacilityUsers can't create new Facilities, regardless of their roles """
+        unsaved_facility = Facility(name="Home")
+        self.assertFalse(unsaved_facility.user_can_create(self.data1["facilityadmin"]))
```

Comment:
It strikes me as a little odd that a Facility must be instantiated in order to check whether a user has *permission* to create it. Specifically, I'm envisioning a scenario where a developer inadvertently persists an object to the database *before* checking whether a user has the permission to create it or not, since instantiating & not saving (`Facility(foo=bar)`) is quite similar to instantiating & saving `Facility.create(foo=bar)`.

What do you think -- is there a way to check permissions that avoids the accidental saving possibility? Or am I splitting hairs?